src/newtondynamics_physics/SysNewton.cpp

```diff
@@ -0,0 +1,428 @@
+    rScene.get_registry().emplace_or_replace<ACompTransformMutable>(ent);
+
+    // Set position/rotation
+    NewtonBodySetMatrix(body, entTransform.m_transform.data());
+
+    // Set damping to 0, as default is 0.1
+    // reference frame may be moving and air pressure stuff
+    NewtonBodySetLinearDamping(body, 0.0f);
+
+    // Make it easier to rotate
+    NewtonBodySetAngularDamping(body, Vector3(1.0f, 1.0f, 1.0f).data());
+
+    // Set callback for updating position of entity and everything else
+    NewtonBodySetForceAndTorqueCallback(body, cb_force_torque);
+
+    // Set user data to entity
+    // WARNING: this will cause issues if 64-bit entities are used on a
+    //          32-bit machine
+    NewtonBodySetUserData(body, reinterpret_cast<void*>(ent));
+}
+
+void SysNewton::compute_rigidbody_inertia(ActiveScene& rScene, ActiveEnt entity)
+{
+    auto& entHier = rScene.reg_get<ACompHierarchy>(entity);
+    auto& entBody = rScene.reg_get<ACompRigidBody>(entity);
+    auto& entNwtBody = rScene.reg_get<ACompNwtBody>(entity);
+
+    // Compute moments of inertia, mass, and center of mass
+    auto const& [inertia, centerOfMass]
+            = SysPhysics::compute_hier_inertia(rScene, entity);
+    entBody.m_centerOfMassOffset = centerOfMass.xyz();
+
+    // Set inertia and mass
+    entBody.m_mass = centerOfMass.w();
+    entBody.m_inertia.x() = inertia[0][0];  // Ixx
+    entBody.m_inertia.y() = inertia[1][1];  // Iyy
+    entBody.m_inertia.z() = inertia[2][2];  // Izz
+
+    NewtonBodySetMassMatrix(entNwtBody.body(), entBody.m_mass,
+        entBody.m_inertia.x(),
+        entBody.m_inertia.y(),
+        entBody.m_inertia.z());
+    NewtonBodySetCentreOfMass(entNwtBody.body(), centerOfMass.xyz().data());
+
+    entBody.m_inertiaDirty = false;
+    SPDLOG_LOGGER_TRACE(m_scene.get_application().get_logger(), "New mass: {}",
+                        entBody.m_mass);
+}
+
+ACompNwtWorld* SysNewton::try_get_physics_world(ActiveScene &rScene)
+{
+    return rScene.get_registry().try_get<ACompNwtWorld>(rScene.hier_get_root());
+}
+
+void SysNewton::on_body_destruct(ActiveReg_t& reg, ActiveEnt ent)
+{
+    NewtonBody const *body = reg.get<ACompNwtBody>(ent).body();
+    NewtonDestroyBody(body); // make sure the Newton body is destroyed
+}
+
+void SysNewton::on_shape_destruct(ActiveReg_t& reg, ActiveEnt ent)
+{
+    NewtonCollision const *shape = reg.get<ACompNwtCollider>(ent).collision();
+    NewtonDestroyCollision(shape); // make sure the shape is destroyed
+}
+
+void SysNewton::on_world_destruct(ActiveReg_t& reg, ActiveEnt ent)
+{
+    NewtonWorld const *world = reg.get<ACompNwtWorld>(ent).m_nwtWorld;
+    if (world != nullptr)
+    {
+        // delete all collision shapes first to prevent crash
+        reg.clear<ACompNwtCollider>();
+
+        // delete all rigid bodies too
+        reg.clear<ACompNwtBody>();
+
+        NewtonDestroy(world); // make sure the world is destroyed
```

Comment:
Can you use an std::unique_ptr with a custom deleter stored in the component to ensure destruction when the component is deleted?

Reply:
It's probably more reliable to use entt destructors here, since components can be moved around often and are fully managed by the registry. Components tend to work better (or just 'more correct') when they're POD.

Reply:
std::unique_ptr vs entt destructors don't make a big different to me.

I just hate to see "manual" resource deletion when some form of pre-registered cleanup logic could be triggered as needed instaed.

Reply:
RAII and data oriented don't go well together. We're also missing out on potential optimizations when the component isn't POD. Ideally, we shouldn't have strings and vectors in the components either.

The entt registry knows when the components are going to be deleted, and it offers this signal as a feature. However, I'd actually want to move this into its own "delete system" instead, as entt happens to delete components polymorphically.